source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -117,115 +185,153 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
-    // This call was invoked during decoding by decoder_callbacks_->encodeHeaders because a fresh
-    // cached response was found and is being added to the encoding stream -- ignore it.
-    return Http::FilterHeadersStatus::Continue;
-  }
+void CacheFilter::UpstreamRequest::abort() {
+  stream_->reset(); // Calls onReset, resulting in deletion.
+}
 
-  // If lookup_ is null, the request wasn't cacheable, so the response isn't either.
-  if (!lookup_) {
-    return Http::FilterHeadersStatus::Continue;
+CacheFilter::UpstreamRequest::~UpstreamRequest() {
+  if (filter_ != nullptr) {
+    filter_->onUpstreamRequestReset();
   }
-
-  if (lookup_result_ == nullptr) {
-    // Filter chain iteration is paused while a lookup is outstanding, but the filter chain manager
-    // can still generate a local reply. One case where this can happen is when a downstream idle
-    // timeout fires, which may mean that the HttpCache isn't correctly setting deadlines on its
-    // asynchronous operations or is otherwise getting stuck.
-    ENVOY_BUG(Http::Utility::getResponseStatus(headers) !=
-                  Envoy::enumToInt(Http::Code::RequestTimeout),
-              "Request timed out while cache lookup was outstanding.");
-    filter_state_ = FilterState::NotServingFromCache;
-    return Http::FilterHeadersStatus::Continue;
+  if (lookup_) {
+    lookup_->onDestroy();
+    lookup_ = nullptr;
   }
+  if (insert_queue_) {
+    // The insert queue may still have actions in flight, so it needs to be allowed
+    // to drain itself before destruction.
+    insert_queue_->setSelfOwned(std::move(insert_queue_));
+  }
+}
 
-  if (filter_state_ == FilterState::ValidatingCachedResponse && isResponseNotModified(headers)) {
-    processSuccessfulValidation(headers);
-    // Stop the encoding stream until the cached response is fetched & added to the encoding stream.
-    if (is_head_request_) {
-      // Return since HEAD requests are not cached
-      return Http::FilterHeadersStatus::Continue;
-    } else {
-      return Http::FilterHeadersStatus::StopIteration;
-    }
+void CacheFilter::UpstreamRequest::onReset() { self_ownership_ = nullptr; }
+void CacheFilter::UpstreamRequest::onComplete() {
+  if (filter_) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::UpstreamRequest complete", *filter_->decoder_callbacks_);
+    filter_->onUpstreamRequestComplete();
+    filter_ = nullptr;
+  } else {
+    ENVOY_LOG(debug, "CacheFilter::UpstreamRequest complete after stream finished");
+  }
+  self_ownership_ = nullptr;
+}
+void CacheFilter::UpstreamRequest::disconnectFilter() {
+  filter_ = nullptr;
+  if (insert_queue_ == nullptr) {
+    abort();
   }
+}
 
+void CacheFilter::UpstreamRequest::onHeaders(Http::ResponseHeaderMapPtr&& headers,
+                                             bool end_stream) {
+  if (filter_state_ == FilterState::ValidatingCachedResponse && isResponseNotModified(*headers)) {
+    return processSuccessfulValidation(std::move(headers));
```

Comment:
I asked about this too, and was told the AsyncHttpClient guarantees that no other callbacks are called after `reset()`. Though hm, I'm not certain if that means `onReset` doesn't get called either, we may need to call it ourselves. Maybe I should validate this behavior in an integration test.

Reply:
Turns out there is already an integration test that covers this path, and I *think* it would complain in one of the *san builds if a memory leak was provoked, which would be the effect if `onReset` didn't get called, so we have a guarantee there. (I also, to be on the safe side, tested in the integration test with some logging to make sure it was behaving as expected.)